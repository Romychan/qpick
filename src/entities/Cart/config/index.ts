/** The name of the storage for storing data in the cart */
export const CART_STORAGE_NAME = 'cart';
