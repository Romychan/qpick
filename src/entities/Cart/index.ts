export { MOCK_CART, MOCK_CART_PRODUCT } from './lib/mock';

export { CartSummary } from './ui/CartSummary/CartSummary';

export { CartContext, CartContextProvider } from './model/context';
export { useCartStore } from './model/hooks';
