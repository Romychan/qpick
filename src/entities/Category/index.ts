export type { Category } from './model/types';
export { MAIN_CATEGORIES } from './model/constants';
