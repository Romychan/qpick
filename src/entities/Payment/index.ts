export { PAYMENT_TYPES } from './model/constants';
export {
  PaymentFormSchema,
  CardPaymentSchema,
  CashPaymentSchema,
} from './model/schemas';
export type {
  PaymentFormType,
  CardPaymentFormType,
  CashPaymentFormType,
} from './model/types';
