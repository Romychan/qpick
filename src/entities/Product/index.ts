export { ProductDetails } from './ui/ProductDetails/ProductDetails';
export { ProductRow } from './ui/ProductRow/ProductRow';
export { ProductCard } from './ui/ProductCard/ProductCard';

export type { Product } from './model/types';

export { MOCK_PRODUCTS } from './lib/mock';
