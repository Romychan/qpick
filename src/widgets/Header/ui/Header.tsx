import { useMemo } from 'react';
import { Link } from 'react-router-dom';

import { useCartStore } from '~/entities/Cart';

import { Logo } from '~/shared/ui/Logo';
import { Badge } from '~/shared/ui/Badge';
import { Icon } from '~/shared/ui/Icon';
import { CART_ROUTE, FAVORITE_ROUTE, MAIN_ROUTE } from '~/shared/config';

import styles from './Header.module.scss';

/**
 * This component is used to display the header of a web page.
 * It contains a logo, a link with a cart and wish list icon.
 */
export const Header = () => {
  const { items: cart } = useCartStore();
  const countCartItems = useMemo(
    () => cart.reduce((acc, cur) => acc + cur.quantity, 0),
    [cart],
  );

  return (
    <header className={styles.container}>
      <div className="container">
        <div className={styles.inner}>
          <Link to={MAIN_ROUTE}>
            <Logo />
          </Link>

          <div className={styles.items}>
            <Link
              to={FAVORITE_ROUTE}
              data-testid="header-favorite"
              aria-label="Перейти на страницу избранным товарам"
              title="Избранное"
              className={styles.link}
            >
              <Badge count={0} colorScheme="yellow">
                <Icon name="heart-filled" size={24} className={styles.icon} />
              </Badge>
            </Link>

            <Link
              to={CART_ROUTE}
              data-testid="header-cart"
              aria-label="Перейти на страницу корины товаров"
              title="Корзина"
              className={styles.link}
            >
              <Badge count={countCartItems} colorScheme="yellow">
                <Icon name="cart-filled" size={24} className={styles.icon} />
              </Badge>
            </Link>
          </div>
        </div>
      </div>
    </header>
  );
};
