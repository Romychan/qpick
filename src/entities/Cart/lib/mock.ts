// eslint-disable-next-line @conarti/feature-sliced/layers-slices
import { Product } from '~/entities/Product/@x';

import { CartItem } from '../model/types';

/** Mock product for the cart */
export const MOCK_CART_PRODUCT: Product = {
  id: 1,
  title: 'Product',
  rate: 4.2,
  image: '/images/catalog/headphone-1.png',
  price: 1200,
  oldPrice: 1300,
  description: 'Test',
  category: 'Наушники',
};

/** List of mock cart items */
export const MOCK_CART: CartItem[] = [
  {
    product: MOCK_CART_PRODUCT,
    quantity: 2,
  },
];
