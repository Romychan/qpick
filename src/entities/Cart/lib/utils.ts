// eslint-disable-next-line @conarti/feature-sliced/layers-slices
import { Product } from '~/entities/Product/@x';

import { CartItem } from '../model/types';

/**
 * Add a product to the cart
 *
 * @param cartItems The current state of the cart
 * @param product Product to add to cart
 * @returns Updated state of the product cart
 */
export const addToCart = (cartItems: CartItem[], product: Product) => {
  const productInCart = cartItems.find(
    (item) => item.product.id === product.id,
  );

  if (!productInCart) {
    return [...cartItems, { product, quantity: 1 }];
  } else {
    return cartItems.map((item) =>
      productInCart.product.id === item.product.id
        ? { ...productInCart, quantity: item.quantity + 1 }
        : item,
    );
  }
};

/**
 * Delete a product from the cart
 *
 * @param cartItems The current state of the cart
 * @param id The ID of the product to be deleted from the cart
 * @returns Updated state of the product cart
 */
export const deleteFromCart = (cartItems: CartItem[], id: number) => {
  return cartItems.filter((item) => item.product.id !== id);
};

/**
 * Increase the number of product in the cart
 *
 * @param cartItems The current state of the cart
 * @param id Product ID to increase the quantity
 * @returns Updated state of the product cart
 */
export const increaseProductQuantity = (cartItems: CartItem[], id: number) => {
  const productInCart = cartItems.find((item) => item.product.id === id);

  if (productInCart) {
    return cartItems.map((item) =>
      productInCart.product.id === item.product.id
        ? { ...productInCart, quantity: item.quantity + 1 }
        : item,
    );
  }

  return cartItems;
};

/**
 * Decrease the number of product in the cart
 *
 * @param cartItems The current state of the cart
 * @param id Product ID to decrease the quantity
 * @returns Updated state of the product cart
 */
export const decreaseProductQuantity = (cartItems: CartItem[], id: number) => {
  const productInCart = cartItems.find((item) => item.product.id === id);

  if (productInCart) {
    return cartItems.map((item) =>
      productInCart.product.id === item.product.id
        ? { ...productInCart, quantity: item.quantity - 1 }
        : item,
    );
  }

  return cartItems;
};
