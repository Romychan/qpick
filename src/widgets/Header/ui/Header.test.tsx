import { RenderResult, screen } from '@testing-library/react';

import { CartContextProvider, MOCK_CART } from '~/entities/Cart';

import { renderWithRouter } from '~/shared/lib/tests';

import { Header } from './Header';

describe('Header', () => {
  let container: RenderResult;

  beforeEach(() => {
    container = renderWithRouter(
      <CartContextProvider preloadState={{ items: MOCK_CART }}>
        <Header />
      </CartContextProvider>,
    );
  });

  it('should render correctly', () => {
    const logoElement = screen.getByTestId('logo');
    const cartElement = screen.getByTestId('header-cart');
    const favoriteElement = screen.getByTestId('header-favorite');

    expect(logoElement).toBeInTheDocument();
    expect(cartElement).toBeInTheDocument();
    expect(cartElement).toHaveTextContent('2');
    expect(favoriteElement).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    expect(container).toMatchSnapshot();
  });
});
