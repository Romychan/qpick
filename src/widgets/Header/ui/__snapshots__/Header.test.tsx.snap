// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="_container_f032e3"
      >
        <div
          class="container"
        >
          <div
            class="_inner_f032e3"
          >
            <a
              href="/"
            >
              <div
                class="_container_4cc90f"
                data-testid="logo"
              >
                <p
                  class="_logo_4cc90f"
                >
                  Qpick
                </p>
              </div>
            </a>
            <div
              class="_items_f032e3"
            >
              <a
                aria-label="Перейти на страницу избранным товарам"
                class="_link_f032e3"
                data-testid="header-favorite"
                href="/favorite"
                title="Избранное"
              >
                <div
                  class="_container_34e0b4"
                >
                  <svg
                    class="icon icon-heart-filled _icon_f032e3 filled"
                    data-testid="icon"
                    height="24"
                    width="24"
                  >
                    <use
                      xlink:href="/images/icons.svg#heart-filled"
                    />
                  </svg>
                </div>
              </a>
              <a
                aria-label="Перейти на страницу корины товаров"
                class="_link_f032e3"
                data-testid="header-cart"
                href="/cart"
                title="Корзина"
              >
                <div
                  class="_container_34e0b4"
                >
                  <svg
                    class="icon icon-cart-filled _icon_f032e3 filled"
                    data-testid="icon"
                    height="24"
                    width="24"
                  >
                    <use
                      xlink:href="/images/icons.svg#cart-filled"
                    />
                  </svg>
                  <span
                    class="_badge_34e0b4 _yellow_34e0b4"
                    data-testid="badge-count"
                  >
                    2
                  </span>
                </div>
              </a>
            </div>
          </div>
        </div>
      </header>
    </div>
  </body>,
  "container": <div>
    <header
      class="_container_f032e3"
    >
      <div
        class="container"
      >
        <div
          class="_inner_f032e3"
        >
          <a
            href="/"
          >
            <div
              class="_container_4cc90f"
              data-testid="logo"
            >
              <p
                class="_logo_4cc90f"
              >
                Qpick
              </p>
            </div>
          </a>
          <div
            class="_items_f032e3"
          >
            <a
              aria-label="Перейти на страницу избранным товарам"
              class="_link_f032e3"
              data-testid="header-favorite"
              href="/favorite"
              title="Избранное"
            >
              <div
                class="_container_34e0b4"
              >
                <svg
                  class="icon icon-heart-filled _icon_f032e3 filled"
                  data-testid="icon"
                  height="24"
                  width="24"
                >
                  <use
                    xlink:href="/images/icons.svg#heart-filled"
                  />
                </svg>
              </div>
            </a>
            <a
              aria-label="Перейти на страницу корины товаров"
              class="_link_f032e3"
              data-testid="header-cart"
              href="/cart"
              title="Корзина"
            >
              <div
                class="_container_34e0b4"
              >
                <svg
                  class="icon icon-cart-filled _icon_f032e3 filled"
                  data-testid="icon"
                  height="24"
                  width="24"
                >
                  <use
                    xlink:href="/images/icons.svg#cart-filled"
                  />
                </svg>
                <span
                  class="_badge_34e0b4 _yellow_34e0b4"
                  data-testid="badge-count"
                >
                  2
                </span>
              </div>
            </a>
          </div>
        </div>
      </div>
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
